feat(security): Enable Authentication and Authorization for Metrics Endpoint using FilterProvider

Zacky3181V opened the pull request:
## Problem Statement

The External Secrets Operator currently exposes the metrics endpoint (:8080/metrics) without any authentication or authorization mechanisms.

## Related Issue

Fixes #5521 

## Proposed Changes

How do you like to solve the issue and why?
Include `metricsOpts.FilterProvider = filters.WithAuthenticationAndAuthorization` to the `/cmd/controller/root.go` inside `if metricsSecure` statement. Also, include `if metricsSecure` statement to the `/cmd/controller/certcontroller.go` and `/cmd/controller/webhook.go` files, since they also expose their own metrics servers 

## Format

Please ensure that your PR follows the following format for the title:
```
feat(scope): add new feature
fix(scope): fix bug
docs(scope): update documentation
chore(scope): update build tool or dependencies
ref(scope): refactor code
clean(scope): provider cleanup
test(scope): add tests
perf(scope): improve performance
desig(scope): improve design
```

Where `scope` is _optionally_ one of:
- charts
- release
- testing
- security
- templating

## Checklist

- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [x] My changes have reasonable test coverage
- [x] All tests pass with `make test`
- [x] I ensured my PR is ready for review with `make reviewable`


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
Enables Kubernetes RBAC-based authentication/authorization for operator metrics endpoints via a new --metrics-auth flag that uses controller-runtime’s filters.WithAuthenticationAndAuthorization as the metrics FilterProvider.

Key changes
- Code
  - Adds --metrics-auth flag and metricsAuth handling in cmd/controller/root.go, cmd/controller/webhook.go, and cmd/controller/certcontroller.go. When enabled, metricsOpts.FilterProvider = filters.WithAuthenticationAndAuthorization.
  - root controller: supports https+auth when --metrics-secure and --metrics-auth are both set. webhook and certcontroller: enable auth (http+auth by default) but TLS was not added to avoid breaking changes.
- Helm charts & manifests
  - New values (default false): metrics.listen.auth.enabled, webhook.metrics.listen.auth.enabled, certController.metrics.listen.auth.enabled.
  - Conditional RBAC rules granting create for authentication.k8s.io/tokenreviews and authorization.k8s.io/subjectaccessreviews when metrics auth is enabled.
  - Conditional service-account-token Secrets per component and ServiceMonitor authorization blocks updated to reference the token secret (Bearer) for Prometheus scraping.
  - Deployment/cert-controller/webhook templates inject --metrics-auth when the corresponding value is enabled.
- Other
  - README, values.yaml, and values.schema.json updated to document the new settings.
  - go.mod updated with additional indirect dependencies.
  - Addresses issue #5521.

Notes
- Metrics authentication is independent from TLS (supports http+auth, https+auth, or no auth). Reviewers suggested optionally extending TLS to webhook and certcontroller in a follow-up for consistency.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Zacky3181V committed:
feat: Enable Authentication and Authorization for Metrics Endpoint using FilterProvider

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
feat(security): Enable Authentication and Authorization for Metrics Endpoint using FilterProvider

This commit reflects requested code changes. Boolean variable metricsAuth created in var() and registered in func init() in root.go to serve as a flag FilterProvider depends on.

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V commented on cmd/controller/webhook.go:
```diff
@@ -127,6 +128,14 @@ var webhookCmd = &cobra.Command{
 			BindAddress: metricsAddr,
 		}
 
+		if metricsSecure {
```
Got it. I've created `metricsAuth` boolean variable in `var()` in `root.go`. Now `FilterProvider` depends on `metricsAuth` flag. The flag is used in all three (root, certcontroller, webhook). 

One question here, should I remove `if metricsSecure` from `certcontroller.go` and `webhook.go`? Currently, when `metricsSecure` is enabled it applies only on `root.go` only. I've included `if metricsSecure` to the `webhook.go` and `certcontroller.go` on the first commit, but after review I am not sure whether it should be present.

Zacky3181V committed:
feature(security): Enable Authentication and Authorization for Metrics Endpoint using FilterProvider

Removed the metricsSecure flag usage from certcontroller and webhook.

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
added metricsSecure to certcontroller and webhook

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
added a new value metrics.listen.auth.enabled
added --metrics-auth in deployment
updated README and values.schema.json with new value

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
created secret with serviceaccount token for each controller

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
extended rbac for all three controllers

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
updated serviceMonitor manifests to use credentials param

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

bharath-b-rh commented on deploy/charts/external-secrets/templates/cert-controller-rbac.yaml:
```diff
@@ -76,6 +76,20 @@ rules:
+    verbs:
+    - "create"
+  - apiGroups:
+    - "authentication.k8s.io"
```
```suggestion
    - "authorization.k8s.io"
```

bharath-b-rh reviewed (commented):
Thank you for considering the suggestions! LGTM

Zacky3181V committed:
added value description in charts' readme
fixed type in rbac for all three controllers

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

bharath-b-rh commented on cmd/controller/certcontroller.go:
```diff
@@ -79,6 +80,15 @@ var certcontrollerCmd = &cobra.Command{
+			metricsServerOpts.CertName = metricsCertName
+			metricsServerOpts.KeyName = metricsKeyName
+		}
+		if metricsAuth {
```
We also need to add the flag in `init()`
```
certcontrollerCmd.Flags().BoolVar(&metricsAuth, "metrics-auth", false, "Enable Kubernetes RBAC-based authentication and authorization for the metrics endpoint.")
```

Zacky3181V replied:
added

bharath-b-rh commented on deploy/charts/external-secrets/templates/servicemonitor.yaml:
```diff
@@ -18,6 +18,12 @@ spec:
+    credentials:
+      name: {{ include "external-secrets.serviceAccountName" . }}-token
+      key: token
+  {{- end }}
```
There is an indentation issue. And also I think we should add this at the end after `relabelings` param. 
Also when `.Values.metrics.listen.auth.enabled` is `true` the change here adds multiple entries instead of updating.
```suggestion
+    {{- if .Values.metrics.listen.auth.enabled }}
+    authorization:
+      credentials:
+        name: {{ include "external-secrets-cert-controller.serviceAccountName" . }}-token
+        key: token
+      type: Bearer
+    {{- end }}
```

Zacky3181V replied:
fixed

bharath-b-rh reviewed (commented):
LGTM, except for the few required modifications. I was able to test this and get the metrics scraping to work on a openshift cluster.

Zacky3181V committed:
Update deploy/charts/external-secrets/templates/servicemonitor.yaml

Co-authored-by: Bharath B <bhb@redhat.com>
Signed-off-by: Daniyar Alkulov <93855417+Zacky3181V@users.noreply.github.com>

Zacky3181V committed:
added metricsAuth flag to the init() function in certcontroller and webhook

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
added if statement to enable metricsAuth flag in the webhook-deployment and certcontroller-deployment yaml files

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
created certController.metrics.listen.auth.enabled and webhook.metrics.listen.auth.enabled to the

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
changed reference from .Values.metrics.listen.auth.enabled to webhook's and certController's own metrics.listen.auth.enabled value

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
updated if statement for the secret with "and" and included values from service account's if statement to make sure that the secret is created when service account is created too

added namespace to the secret

updated reference from .Values.metrics.listen.auth.enabled to certController's and webhook's own metrics.listen.auth.enabled

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
updated reference from .Values.metrics.listen.auth.enabled to certController's and webhook's own in rbac yaml files

fixed indentation in service monitor for root, certController and webhook metrics

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
added namespace for the serviceaccount

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
fixed indentation issue in the service monitor

fixed wrong serviceaccount name in servicemonitor for the webhook and root by referencing their serviceaccounts instead of the certcontroller.

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
added labels to the cert-controller-serviceaccount secret

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
added missing whitespace in servicemonitor

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
deleting whitespaces

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
committed resutls from the make reviewable

added description in README to the webhook.metrics.listen.auth.enabled and certcontroller.metrics.listen.auth.enabled

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
updated go.sum after make reviewable

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

bharath-b-rh commented:
LGTM, Thank you!

Skarlso committed:
make check-diff

Signed-off-by: Gergely Brautigam <182850+Skarlso@users.noreply.github.com>

On-behalf-of: Gergely Brautigam <gergely.brautigam@sap.com>

Skarlso committed:
removed trailing space

Signed-off-by: Gergely Brautigam <182850+Skarlso@users.noreply.github.com>

Zacky3181V committed:
added documentation to readme and values.yaml for auth.enabled

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Skarlso committed:
Merge branch 'main' into main

Skarlso committed:
Merge branch 'main' into main

Signed-off-by: Gergely Bräutigam <skarlso777@gmail.com>

Skarlso committed:
go mod tidy

Signed-off-by: Gergely Brautigam <182850+Skarlso@users.noreply.github.com>

Zacky3181V committed:
deleted metricsSecure from certcontroller and webhook

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
added tests for webhook serviceaccount

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
updated webhook_serviceaccount_test.yaml with all false case

added cert_controller_serviceaccount_test.yaml and serviceaccount_test.yaml covering secret creation with tests

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
added annotations and labels to serviceaccount's secrets

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
updated rbac name to prevent name collision

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
Merge branch 'main' into main

Signed-off-by: Daniyar Alkulov <93855417+Zacky3181V@users.noreply.github.com>

Zacky3181V committed:
check-diff

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Skarlso committed:
Merge branch 'main' into main

Skarlso commented on cmd/controller/certcontroller.go:
```diff
@@ -79,6 +80,10 @@ var certcontrollerCmd = &cobra.Command{
 			BindAddress: metricsAddr,
 		}
 
+		if metricsAuth {
```
This will enable the auth, but secure false will result in bearer token in plain http. That's not desired.

You need to mirror what root.go is doing for secure auth and then add this:

```go
  if metricsAuth && !metricsSecure {
      setupLog.Error(nil, "--metrics-auth requires --metrics-secure; bearer tokens over plaintext HTTP is not allowed")
      os.Exit(1)
  }
```

Zacky3181V replied:
Does mirror what root.go is doing for secure auth mean adding the following code block to the certcontroller.go and webhook.go?
```
		if metricsSecure {
			metricsOpts.SecureServing = true
			metricsOpts.CertDir = metricsCertDir
			metricsOpts.CertName = metricsCertName
			metricsOpts.KeyName = metricsKeyName
		}
```

then adding the code block you suggested to all three controllers (root, certcontroller, webhook)
```
  if metricsAuth && !metricsSecure {
      setupLog.Error(nil, "--metrics-auth requires --metrics-secure; bearer tokens over plaintext HTTP is not allowed")
      os.Exit(1)
  }
```

Is that correct?

Skarlso replied:
Correct, yeah :)

Skarlso commented on cmd/controller/root.go:
```diff
@@ -155,6 +157,9 @@ var rootCmd = &cobra.Command{
 			metricsOpts.CertName = metricsCertName
 			metricsOpts.KeyName = metricsKeyName
 		}
+		if metricsAuth {
+			metricsOpts.FilterProvider = filters.WithAuthenticationAndAuthorization
+		}
```
Change this to the above guard.

Skarlso commented on cmd/controller/webhook.go:
```diff
@@ -127,6 +128,10 @@ var webhookCmd = &cobra.Command{
 			BindAddress: metricsAddr,
 		}
 
+		if metricsAuth {
```
This will enable the auth, but secure false will result in bearer token in plain http. That's not desired.

You need to mirror what root.go is doing for secure auth and then add this:

```go
  if metricsAuth && !metricsSecure {
      setupLog.Error(nil, "--metrics-auth requires --metrics-secure; bearer tokens over plaintext HTTP is not allowed")
      os.Exit(1)
  }
```

Skarlso commented on deploy/charts/external-secrets/templates/webhook-rbac.yaml:
```diff
@@ -0,0 +1,36 @@
+{{- if .Values.webhook.metrics.listen.auth.enabled }}
```
You also need `webhook.create` and `webhook.serviceAccount.create` here.

Skarlso commented on deploy/charts/external-secrets/values.yaml:
```diff
@@ -321,6 +321,9 @@ metrics:
 
   listen:
     port: 8080
+    auth:
```
document that the default is false please. The others too. :)

Skarlso reviewed (commented):
Almost there. :)

Zacky3181V committed:
updated conditional argument in the webhook-rbac to include .Values.webhook.create .Values.webhook.serviceAccount.create

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
added text that the default auth's value is false.

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Zacky3181V committed:
extended metricsSecure to the webhook and certController.

Signed-off-by: Daniyar Alkulov <medvedicabolsaa@gmail.com>

Skarlso committed:
Merge branch 'main' into main

Signed-off-by: Gergely Bräutigam <gergely.brautigam@sap.com>

Skarlso committed:
go mod tidy

Signed-off-by: Gergely Brautigam <182850+Skarlso@users.noreply.github.com>

Skarlso commented:
/ok-to-test sha=44d36c9c9fede0297c33ba9080f08af3c516dbed

Skarlso commented:
/ok-to-test sha=d07761c6556cc2181082352fb619cbaf49865825

Skarlso reviewed (changes requested):
Ah I found some tiny bugs, I'll fix them in a sec.

Skarlso committed:
put back the service account annotation and use the right scope

Signed-off-by: Gergely Brautigam <182850+Skarlso@users.noreply.github.com>

Skarlso reviewed (approved):
Now, it's okay, I think.

Skarlso commented:
/ok-to-test sha=134844cb47bec870f85b5a9254a1eee87387a8b7
